### DISCUSSION TITLE
Ignore Inserts/Updates/Deletes feature

### DISCUSSION THREAD

**Discussion by JiriOgurek** [Mimetis/Dotmim.Sync#669] [Ideas]
As an addendum to SyncDirection (Download/UploadOnly) I would suggest to add ignore Inserts/Updates/Deletes feature to cover all possible sync scenarios, such as mine ;-) What do you think about it?

**Comment by Mimetis** [comment]
Can you explain more what kind of scenario you have in mind ?

**Comment by JiriOgurek** [comment]
Sure, to be able ignore (to not synchronize) new inserted/updated/deleted records. It means to be able to do not track these (I/U/D) changes, or effeciantely delete them prior to sync.

**Reply by Mimetis** [reply]
If you are not syncing any insert update delete... Well there is nothing to sync, right ?

**Reply by JiriOgurek** [reply]
Ignore Inserts/Updates/Deletes, "/" sign means OR. In my scenario, clients provide new data only and central server collects it, I am interested in Inserts only, but I can imagine other useful scenarios, this would be very handy and easy to use feature.

**Reply by Mimetis** [reply]
There is no distinction between `Insert` or `Update`. It's kind of "`Upsert`"

If you want to avoid any **deleted row** to be sync to the server, you still have a chance to do it, from the **client side** (using the `LocalOrchestrator`), that way:

``` cs
 agent.LocalOrchestrator.OnTableChangesSelectedSyncRow(args =>
 {
     if (args.SyncRow.RowState == DataRowState.Deleted)
         args.SyncRow = null;
 });
```

If you want to allow `Insert` and disable `Update`, it will be a little bit more complex.
Basically, you can't know from the client side if the row is new or already existing on the server side.

So far, you need to intercept the row on the **server** and see if it already exists, using the `RemoteOrchestrator` (or `WebServerOrchestrator` if you are being a web api), that way:

``` cs
agent.RemoteOrchestrator.OnTableChangesApplyingSyncRows(async args =>
{
    if (args.SchemaTable.TableName == "ProductCategory")
    {
        var cmd = args.Connection.CreateCommand();
        cmd.CommandText = "Select count(*) from ProductCategory_tracking where ProductCategoryID = @ProductCategoryID";
        cmd.Connection = args.Connection;
        cmd.Transaction = args.Transaction;
        var p = cmd.CreateParameter();
        p.DbType = DbType.Guid;
        p.ParameterName = "@ProductCategoryID";
        cmd.Parameters.Add(p);

        foreach (var row in args.SyncRows.ToArray())
        {
            if (row.RowState == DataRowState.Modified)
            {
                cmd.Parameters[0].Value = new Guid(row["ProductCategoryID"].ToString());
                var alreadyExists = await cmd.ExecuteScalarAsync();

                if ((int)alreadyExists == 1)
                {
                    args.SyncRows.Remove(row);
                }
            }
        }

    }
});
```

It's specific to the ProducCategory table, but you can find a way to generate any sql for any table, of course

**Reply by JiriOgurek** [reply]
It's been proposed to use OnTableChangesSelected event for deletes removal, in post above you have mentioned OnTableChangesSelectedSyncRow event, can you explain the difference of these two events, please?

**Reply by Mimetis** [reply]
`OnTableChangesSelected` is raised after all rows have been selected, and these rows are already seriliazed on disk and ready to be sent

`OnTableChangesSelectedSyncRow` is raised after during reading the data source. The row is still in memory and is not been, yet, serialized on the disk.

For your scenario, it's just a performance trick to avoid reading multiple times the row

**Reply by JiriOgurek** [reply]
So SyncRow event is prior to the other one, right?

**Reply by Mimetis** [reply]
yes, you can see it in the code here https://github.com/Mimetis/Dotmim.Sync/blob/c69122b77bc509dfa36344945b2ffc1e35deb1f7/Projects/Dotmim.Sync.Core/Orchestrators/BaseOrchestrator.GetChanges.cs#L158

**Reply by JiriOgurek** [reply]
Thank you for explanation, this leads me to initial idea to have boolean Inserts/Updates/Deletes Interest feature (default = true) to subscribe wanted changes, this would be very user-friendly feature.
